### PR TITLE
Return `_ignored_source` only if explicitly requested via `stored_fields` or `fields`

### PR DESCRIPTION
**Pull request by salvatore-campagna**
We do not want to return `_ignored_source` with every search hit. Other than being expensive
(especially when there are many ignored fields), it also exposes some implementation details. Anyway, it might
still be useful to have the ability to retrieve it if necessary, at least for debugging purposes. For this
reason we require that it is explicitly requested using `stored_fields` or `fields` and we do not return
it by default or if requested via a wildcard `*` in `stored_fields` or `fields`.

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/RestTestBasePlugin.java [hunk 1]
```diff
@@ -169,7 +169,7 @@ public void apply(Project project) {
             nonInputSystemProperties.systemProperty(TESTS_MAX_PARALLEL_FORKS_SYSPROP, () -> String.valueOf(task.getMaxParallelForks()));
 
             // Disable test failure reporting since this stuff is now captured in build scans
-            task.getExtensions().getByType(ErrorReportingTestListener.class).setDumpOutputOnFailure(false);
```
Comment (salvatore-campagna) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/RestTestBasePlugin.java:172]: sorry will remove this

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/560_ignored_source.yml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+          mappings:
+            _source:
+              mode: synthetic
+            properties:
+              object:
+                enabled: false
+
+  - do:
+      headers:
+        Content-Type: application/yaml
```
Comment (salvatore-campagna) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/560_ignored_source.yml:18]: I had to include this because it looks like sometimes the source in in json and sometimes is in yaml. This changes the base64 encoded value I check later in `match`.

File: server/src/main/java/org/elasticsearch/search/SearchHit.java [hunk 3]
```diff
@@ -853,6 +854,8 @@ public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) t
             // TODO: can we avoid having an exception here?
             if (field.getName().equals(IgnoredFieldMapper.NAME)) {
                 builder.field(field.getName(), field.getValues());
+            } else if (field.getName().equals(IgnoredSourceFieldMapper.NAME)) {
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/search/SearchHit.java:857]: I think this is not needed here. Yaml tests produce output as bytes but that is likely a yaml test issue. I will remove it.

File: server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java [hunk 4]
```diff
@@ -315,4 +319,17 @@ public static byte[] encodeFromMap(MappedNameValue mappedNameValue, Map<String,
...
+
+    /**
+     * Encodes the provided {@link BytesRef} into a Base64 string representation.
+     *
+     * @param value the {@link BytesRef} to be encoded. This must not be null.
+     * @return a {@link String} representing the Base64 encoded version of the bytes in the provided {@link BytesRef}.
+     *
+     * @throws NullPointerException if the provided {@link BytesRef} is null.
+     */
+    public static String base64Encode(final BytesRef value) {
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java:331]: This is not needed anymore.

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java [hunk 5]
```diff
@@ -97,6 +101,10 @@ protected void doAssert(Object actualValue, Object expectedValue) {
             assertThat(actualValue, instanceOf(List.class));
             assertMap((List<?>) actualValue, matchesList((List<?>) expectedValue));
         }
+        if (expectedValue instanceof String expectedValueAsString && actualValue instanceof byte[] actualValueAsBytes) {
```
Comment (salvatore-campagna) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java:104]: I had to add this here in order to be able to compare a byte array (like `_ignored_source`) using Base64 encoding. I don't see any other way to compare array of bytes.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java [hunk 6]
```diff
@@ -56,23 +58,47 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
         }
 
         final SearchExecutionContext searchExecutionContext = fetchContext.getSearchExecutionContext();
-        final FieldFetcher fieldFetcher = fetchFieldsContext == null ? null
-            : fetchFieldsContext.fields() == null ? null
-            : fetchFieldsContext.fields().isEmpty() ? null
-            : FieldFetcher.create(searchExecutionContext, fetchFieldsContext.fields());
+        final FieldFetcher fieldFetcher;
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:61]: This is required so that we can fetch `_ignored_source` via `fields`

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java [hunk 7]
```diff
@@ -56,23 +58,47 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
...
             && storedFieldsContext.fieldNames().isEmpty() == false) {
-            final Set<FieldAndFormat> metadataFields = new HashSet<>(DEFAULT_METADATA_FIELDS);
             for (final String storedField : storedFieldsContext.fieldNames()) {
                 final Set<String> matchingFieldNames = searchExecutionContext.getMatchingFieldNames(storedField);
                 for (final String matchingFieldName : matchingFieldNames) {
                     if (SourceFieldMapper.NAME.equals(matchingFieldName) || IdFieldMapper.NAME.equals(matchingFieldName)) {
                         continue;
                     }
                     final MappedFieldType fieldType = searchExecutionContext.getFieldType(matchingFieldName);
+                    if (Regex.isSimpleMatchPattern(storedField) && IgnoredSourceFieldMapper.NAME.equals(matchingFieldName)) {
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:99]: This is required to avoid fetching `_ignored_source` when requested via a wildcard `*`

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java [hunk 8]
```diff
@@ -56,23 +58,47 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
...
+                    fieldAndFormat -> searchExecutionContext.isMetadataField(fieldAndFormat.field) == false
+                        || searchExecutionContext.getFieldType(fieldAndFormat.field).isStored() == false
+                )
+                .toList();
+            fieldFetcher = FieldFetcher.create(searchExecutionContext, fetchFields);
+        } else {
+            fieldFetcher = FieldFetcher.create(searchExecutionContext, Collections.emptyList());
+        }
 
+        final Set<FieldAndFormat> additionalMetadataFields = new HashSet<>();
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:75]: This is required to:
* fetch default metadata fields **always**
* fetch metadata stored fields requested via `fields`

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java [hunk 9]
```diff
@@ -61,6 +63,19 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
             : fetchFieldsContext.fields().isEmpty() ? null
             : FieldFetcher.create(searchExecutionContext, fetchFieldsContext.fields());
 
+        final Set<FieldAndFormat> fetchContextMetadataFields = new HashSet<>();
+        if (fetchFieldsContext != null && fetchFieldsContext.fields() != null && fetchFieldsContext.fields().isEmpty() == false) {
+            for (final FieldAndFormat fieldAndFormat : fetchFieldsContext.fields()) {
+                if (SourceFieldMapper.NAME.equals(fieldAndFormat.field) || IdFieldMapper.NAME.equals(fieldAndFormat.field)) {
```
Comment (lkts) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:69]: I don't understand why source and id are involved here. Is this a different existing problem that they are returned in cases when they shouldn't be?

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java (hunk 5 above)
Comment (lkts) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java:104]: You can use this format in yaml and remove this change (it's pretty obscure feature but i tested it and it works).
```
- match: { hits.hits.0._ignored_source: !!binary "BgAAAG9iamVjdHktLS0KbmFtZTogImZvbyIKdmFsdWU6IDEwCg==" }
```

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/560_ignored_source.yml [hunk 10]
```diff
@@ -0,0 +1,186 @@
+---
+setup:
```
Comment (martijnvg) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/560_ignored_source.yml:2]: maybe move this yaml file to logsdb module? Then we don't have to add yet another node feature.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java (hunk 9 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:69]: @salvatore-campagna maybe add a comment explaining why this is necessary?

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java (hunk 5 above)
Comment (martijnvg) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java:104]: I think we should use the obscure format instead of modifying the yaml assert code.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java [hunk 11]
```diff
@@ -81,9 +99,12 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
                     }
                 }
             }
+            metadataFields.addAll(fetchContextMetadataFields);
             metadataFieldFetcher = FieldFetcher.create(searchExecutionContext, metadataFields);
         } else {
-            metadataFieldFetcher = FieldFetcher.create(searchExecutionContext, DEFAULT_METADATA_FIELDS);
+            final Set<FieldAndFormat> allMetadataFields = new HashSet<>(DEFAULT_METADATA_FIELDS);
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:105]: What is the reason behind this change?

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java (hunk 9 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:69]: `_source` and `_id` are treated differently because they are retrieved by other means and not by this logic...so we skip them to avoid duplicates. You can see similar logic in `StoredFieldsPhase`.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java (hunk 11 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:105]: This is required to be able to fetch some stored metadata fields, like `ignored_source` when tehy are requested via `fields`. Normally such fields would be requested via `stored_fields`. After the latest changes (done like 6 months ago) the decision was to fetch fields using the `ValueFetcher` interface which means we do not distinguish anymore stored versus non-stored. Anyway the idea, as you can see from this code is to use the thise part of the logic to fetch metadata fields. In this case we need to do this to fetch metadata fields requested via `fields` (and not via `stored_fields`).

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java (hunk 9 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:69]: If you look into `FieldVisitor` and how it is used you will see logic like the following

```
public void reset() {
        if (fieldsValues != null) {
            fieldsValues.clear();
        }
        source = null;
        id = null;

        requiredFields.addAll(BASE_REQUIRED_FIELDS);
        if (loadSource) {
            requiredFields.add(sourceFieldName);
        }
    }
```

```
BASE_REQUIRED_FIELDS
```

includes `_source` and `_id`. This is just to say that the responsibility of loading those fields is somewhere else (`StoredFieldLoader`).

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java (hunk 5 above)
Comment (salvatore-campagna) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java:104]: Nice...I didn't know that

**Review by martijnvg — APPROVED** [review]
LGTM - I think this should also be back ported to 8.x branch?

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java (hunk 9 above)
Comment (lkts) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:71]: But how did it work when this code was not here?

**Review by lkts — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java (hunk 9 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:71]: Note that some of the yaml tests I wrote actually work in main.

The ones failing is main are the following ones
```
* fetch stored fields wildcard => in main we return `_ignored_source` when `stored_fields: [*]`
* wildcard fields and stored fields combination => same as before
* fields with ignored source in fields => this one is failing because fields requested via `fields` and so in `FetchFieldContext` are fetched via `ValueFetcher` which uses a stored field but the `FiedlFetcher` for **non-metadata** fields put those values in `fields` and not in `metadata fields`. This is an issue because `_ignored_source` is a metadata field which ends up into doc fields
* fetch fields by name => fails for a reason similar to above
```

For the third test failing above see this

```
public void process(HitContext hitContext) throws IOException {
                final Map<String, DocumentField> fields = fieldFetcher != null
                    ? fieldFetcher.fetch(hitContext.source(), hitContext.docId())
                    : Collections.emptyMap();
                final Map<String, DocumentField> metadataFields = metadataFieldFetcher.fetch(hitContext.source(), hitContext.docId());
                hitContext.hit().addDocumentFields(fields, metadataFields);
            }
```

Note that there is a `fieldFetcher` for document fields and a `metadataFieldFetcher` for metadata fields.

@lkts does it make sense?

Comment (lkts) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:71]: Thank you, i understand now!

**Comment by salvatore-campagna** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/114477)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
